File: quant/dcf.py
#!/usr/bin/env python3
"""
DCF for token fee-capture businesses (L1s, DEXs, stablecoin issuers).

Inputs: projected annual fees/revenue for N years, discount rate, terminal growth, circulating supply.
Outputs: fair value per token + scenario range (bear/base/bull multipliers).

Usage:
    python3 dcf.py --fees-yearly '[120e6,150e6,180e6,200e6,210e6]' \
        --discount 0.15 --terminal-growth 0.04 --circulating-supply 1e9
    # Output includes fair value per token, scenarios, sensitivity to discount rate
"""
import argparse
import json


def dcf(fees_yearly, discount, terminal_growth, circulating_supply, capture_ratio=1.0):
    """
    fees_yearly: list of projected annual fee/revenue in dollars
    discount: annual discount rate (e.g. 0.15)
    terminal_growth: perpetual growth rate after last year (e.g. 0.03)
    circulating_supply: tokens outstanding (not max supply)
    capture_ratio: fraction of fees captured by token holders (burn + staking yield); 1.0 = 100% pass-through
    """
    years = len(fees_yearly)
    pv_fees = 0.0
    for t, fee in enumerate(fees_yearly, start=1):
        pv_fees += (fee * capture_ratio) / ((1 + discount) ** t)

    # terminal value at end of year N, assuming growing perpetuity
    terminal_fee = fees_yearly[-1] * (1 + terminal_growth) * capture_ratio
    if discount > terminal_growth:
        tv = terminal_fee / (discount - terminal_growth)
        pv_tv = tv / ((1 + discount) ** years)
    else:
        pv_tv = 0  # degenerate case
    enterprise_value = pv_fees + pv_tv
    fair_value_per_token = enterprise_value / circulating_supply
    return {
        "pv_of_explicit_fees": round(pv_fees, 2),
        "pv_of_terminal": round(pv_tv, 2),
        "enterprise_value": round(enterprise_value, 2),
        "fair_value_per_token": round(fair_value_per_token, 4),
        "terminal_weight_pct": round(pv_tv / enterprise_value * 100, 2) if enterprise_value > 0 else 0,
    }


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--fees-yearly", required=True, help="JSON array of yearly fees in USD")
    ap.add_argument("--discount", type=float, default=0.15)
    ap.add_argument("--terminal-growth", type=float, default=0.03)
    ap.add_argument("--circulating-supply", type=float, required=True)
    ap.add_argument("--capture-ratio", type=float, default=1.0, help="fraction of fees accruing to token holders (1.0 = 100 pct)")
    ap.add_argument("--current-price", type=float, help="optional: compare to current market price")
    args = ap.parse_args()

    fees = json.loads(args.fees_yearly)

    base = dcf(fees, args.discount, args.terminal_growth, args.circulating_supply, args.capture_ratio)

    # scenarios — bear: -40% fees, higher discount, lower terminal; bull: +40%, lower discount
    bear_fees = [f * 0.6 for f in fees]
    bull_fees = [f * 1.4 for f in fees]
    bear = dcf(bear_fees, args.discount + 0.05, max(0.01, args.terminal_growth - 0.02), args.circulating_supply, args.capture_ratio)
    bull = dcf(bull_fees, max(0.05, args.discount - 0.03), args.terminal_growth + 0.02, args.circulating_supply, args.capture_ratio)

    # sensitivity to discount rate
    sens = {}
    for d in [0.10, 0.12, 0.15, 0.18, 0.20, 0.25]:
        sens[f"discount_{d}"] = round(dcf(fees, d, args.terminal_growth, args.circulating_supply, args.capture_ratio)["fair_value_per_token"], 4)

    out = {
        "inputs": {
            "years": len(fees),
            "discount": args.discount,
            "terminal_growth": args.terminal_growth,
            "circulating_supply": args.circulating_supply,
            "capture_ratio": args.capture_ratio,
        },
        "base_case": base,
        "bear_case": bear,
        "bull_case": bull,
        "sensitivity_per_token": sens,
    }
    if args.current_price:
        out["current_price"] = args.current_price
        out["upside_base_pct"] = round((base["fair_value_per_token"] / args.current_price - 1) * 100, 2)
        out["upside_bull_pct"] = round((bull["fair_value_per_token"] / args.current_price - 1) * 100, 2)
        out["downside_bear_pct"] = round((bear["fair_value_per_token"] / args.current_price - 1) * 100, 2)

    print(json.dumps(out, indent=2))


if __name__ == "__main__":
    main()

File: quant-jvm/src/main/kotlin/org/maxmoran/quant/Dcf.kt
package org.maxmoran.quant

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlin.math.max
import kotlin.math.pow
import kotlin.system.exitProcess

private val PRETTY_JSON = Json { prettyPrint = true }

/**
 * Deterministic port of `quant/dcf.py` — token fee-capture DCF with scenario range and
 * discount-rate sensitivity. Plain §1.1 arithmetic (no solver, no RNG).
 *
 * Serialization note: dollar-scale magnitudes (>= 1e7) print in scientific notation on the
 * JVM where Python prints plain decimals. The numbers are identical; JSON text form is
 * non-contract per parity-contract.md §2.
 */

/** Unrounded valuation state; [degenerateTerminal] marks discount <= terminal growth. */
data class DcfResult(
    val pvFees: Double,
    val pvTerminal: Double,
    val degenerateTerminal: Boolean,
    val enterpriseValue: Double,
    val fairValuePerToken: Double,
)

/** Present value of explicit fees plus growing-perpetuity terminal, mirroring `dcf`. */
fun dcfValuation(
    feesYearly: List<Double>,
    discount: Double,
    terminalGrowth: Double,
    circulatingSupply: Double,
    captureRatio: Double = 1.0,
): DcfResult {
    val years = feesYearly.size
    var pvFees = 0.0
    for (t in 1..years) {
        pvFees += (feesYearly[t - 1] * captureRatio) / (1.0 + discount).pow(t)
    }
    val terminalFee = feesYearly.last() * (1.0 + terminalGrowth) * captureRatio
    val degenerate = !(discount > terminalGrowth)
    val pvTerminal = if (degenerate) {
        0.0
    } else {
        val tv = terminalFee / (discount - terminalGrowth)
        tv / (1.0 + discount).pow(years)
    }
    val enterpriseValue = pvFees + pvTerminal
    return DcfResult(
        pvFees = pvFees,
        pvTerminal = pvTerminal,
        degenerateTerminal = degenerate,
        enterpriseValue = enterpriseValue,
        fairValuePerToken = enterpriseValue / circulatingSupply,
    )
}

/**
 * Rounded JSON block for one valuation, field-for-field with the Python `dcf` dict.
 * In the degenerate branch Python's `round(0, 2)` keeps the int, hence the bare 0 here;
 * same for the `else 0` arm of the terminal weight.
 */
fun dcfJson(result: DcfResult): JsonObject = buildJsonObject {
    put("pv_of_explicit_fees", JsonPrimitive(round2(result.pvFees)))
    put(
        "pv_of_terminal",
        if (result.degenerateTerminal) JsonPrimitive(0) else JsonPrimitive(round2(result.pvTerminal)),
    )
    put("enterprise_value", JsonPrimitive(round2(result.enterpriseValue)))
    put("fair_value_per_token", JsonPrimitive(round4(result.fairValuePerToken)))
    put(
        "terminal_weight_pct",
        if (result.enterpriseValue > 0.0) {
            JsonPrimitive(round2(result.pvTerminal / result.enterpriseValue * 100.0))
        } else {
            JsonPrimitive(0)
        },
    )
}

private val SENSITIVITY_DISCOUNTS = listOf(0.10, 0.12, 0.15, 0.18, 0.20, 0.25)

/** Testable result of evaluating the DCF CLI contract. */
data class DcfCliResult(val exitCode: Int, val output: String)

/** Parse input and produce CLI output without terminating the process. */
fun evaluateDcf(args: Array<String>): DcfCliResult {
    return try {
        var feesJson: String? = null
        var discount = 0.15
        var terminalGrowth = 0.03
        var circulatingSupply: Double? = null
        var captureRatio = 1.0
        var currentPrice: Double? = null
        var index = 0

        fun optionValue(option: String): String {
            if (index + 1 >= args.size || args[index + 1].startsWith("--")) {
                throw IllegalArgumentException("$option requires a value")
            }
            index += 1
            return args[index]
        }

        while (index < args.size) {
            when (val arg = args[index]) {
                "--fees-yearly" -> feesJson = optionValue(arg)
                "--discount" -> discount = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--discount must be a number")
                "--terminal-growth" -> terminalGrowth = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--terminal-growth must be a number")
                "--circulating-supply" -> circulatingSupply = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--circulating-supply must be a number")
                "--capture-ratio" -> captureRatio = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--capture-ratio must be a number")
                "--current-price" -> currentPrice = optionValue(arg).toDoubleOrNull()
                    ?: throw IllegalArgumentException("--current-price must be a number")
                else -> throw IllegalArgumentException("unknown argument: $arg")
            }
            index += 1
        }

        // --fees-yearly is inline JSON on the command line, not a file path.
        val fees = parseNumberArray(
            feesJson ?: throw IllegalArgumentException("need --fees-yearly"),
            "fees",
        )
        val supply = circulatingSupply ?: throw IllegalArgumentException("need --circulating-supply")

        val base = dcfValuation(fees, discount, terminalGrowth, supply, captureRatio)
        val bear = dcfValuation(
            fees.map { it * 0.6 },
            discount + 0.05,
            max(0.01, terminalGrowth - 0.02),
            supply,
            captureRatio,
        )
        val bull = dcfValuation(
            fees.map { it * 1.4 },
            max(0.05, discount - 0.03),
            terminalGrowth + 0.02,
            supply,
            captureRatio,
        )

        val output = buildJsonObject {
            put("inputs", buildJsonObject {
                put("years", JsonPrimitive(fees.size))
                put("discount", JsonPrimitive(discount))
                put("terminal_growth", JsonPrimitive(terminalGrowth))
                put("circulating_supply", JsonPrimitive(supply))
                put("capture_ratio", JsonPrimitive(captureRatio))
            })
            put("base_case", dcfJson(base))
            put("bear_case", dcfJson(bear))
            put("bull_case", dcfJson(bull))
            put("sensitivity_per_token", buildJsonObject {
                for (d in SENSITIVITY_DISCOUNTS) {
                    // Key mirrors Python's f"discount_{d}" repr (0.20 prints as 0.2).
                    put(
                        "discount_$d",
                        JsonPrimitive(round4(round4(
                            dcfValuation(fees, d, terminalGrowth, supply, captureRatio).fairValuePerToken
                        ))),
                    )
                }
            })
            // Python gates on truthiness, so an explicit 0 price suppresses the block.
            if (currentPrice != null && currentPrice != 0.0) {
                val price = currentPrice
                put("current_price", JsonPrimitive(price))
                // Upside uses the already-rounded fair values, exactly like the reference.
                put("upside_base_pct", upsidePct(round4(base.fairValuePerToken), price))
                put("upside_bull_pct", upsidePct(round4(bull.fairValuePerToken), price))
                put("downside_bear_pct", upsidePct(round4(bear.fairValuePerToken), price))
            }
        }
        DcfCliResult(
            exitCode = 0,
            output = PRETTY_JSON.encodeToString(JsonObject.serializer(), output),
        )
    } catch (exception: Exception) {
        val message = exception.message?.lineSequence()?.firstOrNull()?.take(240) ?: "invalid input"
        DcfCliResult(
            exitCode = 1,
            output = buildJsonObject { put("error", JsonPrimitive(message)) }.toString(),
        )
    }
}

private fun upsidePct(roundedFairValue: Double, price: Double): JsonPrimitive =
    JsonPrimitive(round2((roundedFairValue / price - 1.0) * 100.0))

/** Module CLI runner — JSON in / JSON out, matching the Python entrypoint. */
fun runDcf(args: Array<String>) {
    val result = evaluateDcf(args)
    println(result.output)
    if (result.exitCode != 0) exitProcess(result.exitCode)
}
